source/opt/local_ssa_elim_pass.cpp

```diff
@@ -0,0 +1,759 @@
+        differs = true;
+        break;
+      }
+      if (var_val_itr->second != val0Id) {
+        differs = true;
+        break;
+      }
+    }
+    // If val is the same for all predecessors, enter it in map
+    if (!differs) {
+      label2ssa_map_[label].insert(var_val);
+      continue;
+    }
+    // Val differs across predecessors. Add phi op to block and 
+    // add its result id to the map
+    std::vector<ir::Operand> phi_in_operands;
+    const uint32_t typeId = GetPointeeTypeId(def_use_mgr_->GetDef(varId));
+    for (uint32_t predLabel : label2preds_[label]) {
+      const auto var_val_itr = label2ssa_map_[predLabel].find(varId);
+      // If variable not defined on this path, use undef
+      const uint32_t valId = (var_val_itr != label2ssa_map_[predLabel].end()) ?
+          var_val_itr->second : Type2Undef(typeId);
+      phi_in_operands.push_back(
+          {spv_operand_type_t::SPV_OPERAND_TYPE_ID, {valId}});
+      phi_in_operands.push_back(
+          {spv_operand_type_t::SPV_OPERAND_TYPE_ID, {predLabel}});
+    }
+    const uint32_t phiId = TakeNextId();
+    std::unique_ptr<ir::Instruction> newPhi(
+      new ir::Instruction(SpvOpPhi, typeId, phiId, phi_in_operands));
+    def_use_mgr_->AnalyzeInstDefUse(&*newPhi);
+    insertItr = insertItr.InsertBefore(std::move(newPhi));
+    ++insertItr;
+    label2ssa_map_[label].insert({varId, phiId});
+  }
+}
+
+bool LocalSSAElimPass::IsLoopHeader(ir::BasicBlock* block_ptr) const {
+  auto iItr = block_ptr->end();
+  --iItr;
+  if (iItr == block_ptr->begin())
+    return false;
+  --iItr;
+  return iItr->opcode() == SpvOpLoopMerge;
+}
+
+void LocalSSAElimPass::SSABlockInit(
+    std::list<ir::BasicBlock*>::iterator block_itr) {
+  const size_t numPreds = label2preds_[(*block_itr)->id()].size();
+  if (numPreds == 0)
+    return;
+  if (numPreds == 1)
+    SSABlockInitSinglePred(*block_itr);
+  else if (IsLoopHeader(*block_itr))
+    SSABlockInitLoopHeader(block_itr);
+  else
+    SSABlockInitMultiPred(*block_itr);
+}
+
+void LocalSSAElimPass::PatchPhis(uint32_t header_id, uint32_t back_id) {
+  ir::BasicBlock* header = id2block_[header_id];
+  auto phiItr = header->begin();
+  for (; phiItr->opcode() == SpvOpPhi; ++phiItr) {
+    uint32_t cnt = 0;
+    uint32_t idx;
+    phiItr->ForEachInId([&cnt,&back_id,&idx](uint32_t* iid) {
+      if (cnt % 2 == 1 && *iid == back_id) idx = cnt - 1;
+      ++cnt;
+    });
+    // Only patch operands that are in the backedge predecessor map
+    const uint32_t varId = phiItr->GetSingleWordInOperand(idx);
+    const auto valItr = label2ssa_map_[back_id].find(varId);
+    if (valItr != label2ssa_map_[back_id].end()) {
+      phiItr->SetInOperand(idx, { valItr->second });
+      // Analyze uses now that they are complete
+      def_use_mgr_->AnalyzeInstUse(&*phiItr);
+    }
+  }
+}
+
+bool LocalSSAElimPass::LocalSSAElim(ir::Function* func) {
+  InitSSARewrite(*func);
+  // Process all blocks in structured order. 
```

Comment:
It is just one way (the simplest?) to make sure that all predecessor blocks are processed before a block itself.